add an option to explicitly disable query IPv6 AAAA, SRV, MX, etc...

muhammadn opened the pull request:
Based on this patch i had submitted to RedHat (https://issues.redhat.com/browse/RHEL-56280), i am submitting this patch to this upstream systemd.

There is no way to explicitly enable/disable IPv6 AAAA queries.

Problem was that i am using RHEL9 and some applications does not use a newer glibc that supports `no-aaaa` option in `/etc/resolv.conf`. So some applications will still resolve IPv6 AAAA even with `no-aaaa` option and it is inconsistent across the system where some work and some don't.

So this systemd-resolved patch catch-all queries and disable IPv6 AAAA queries for all applications in the OS by having an option `DisableAAAAQuery=yes` to disable IPv6 AAAA queries.

Although https://github.com/systemd/systemd/pull/28136 tries to fix this automatically but it still does not work with `net.ipv6.conf.all.disable_ipv6 = 1`. Also tried with explicitly removing the conditional and force set `family = AF_INET` and still resolves AAAA records.

The issue is that i want to explicitly disable IPv6 AAAA queries instead of systemd-resolved to figure out itself which address family it is using, which always have problems.

poettering commented:
Uh, this sounds awfully specific.

I'd be fine if there was an option RefuseRecordType= or so, which takes a list of RR types. That way you could do "RefuseRecordType=AAAA" to do what you want.

poettering commented on src/resolve/dns-type.c:
```diff
@@ -104,6 +104,12 @@ bool dns_class_is_valid_rr(uint16_t class) {
+bool dns_type_is_aaaa_rr(uint16_t type) {
+              /* Check if it is IPv6 AAAA resource record */
+
+              return type == DNS_TYPE_AAAA;
```
indentation is off

muhammadn replied:
I'll take note of this in the future.

poettering commented on src/resolve/resolved-manager.h:
```diff
@@ -141,6 +141,9 @@ struct Manager {
         bool read_etc_hosts;
 
+        /* Enable/Disable IPv6 AAAA record query */
+        bool disable_aaaa_query;
```
negative options are frowned upon in our codebase. i.e. options that literally you need to *enable* to *disable* a feature is just weird.

muhammadn commented:
-@poettering I am doing some reworking to do `RefuseRecordType=` option.

But i cannot wrap around my head as i cannot find any utility function that converts strings like `AAAA` to uint16_t like DNS_TYPE_AAAA. 

I wanted to write an enum that maps to "AAAA" string , initializes a uint16_t variable and sets that variable value to DNS_TYPE_AAA enum that but it would be messy so i think i would need to get some ideas around this.

muhammadn commented:
Just found `DEFINE_STRING_TABLE_LOOKUP`. I'll take a look at it.

poettering commented:
there already is dns_type_from_string()

poettering commented:
and please parse the RR types when parsing configuration, and store them internally in a `Set` object in parsed (i.e. numeric) form, so that lookups are reasonably quick.

muhammadn commented:
@poettering I'll add a new parser because of some parts i have to think about in `conf-parser.c` to use the `dns_type_from_string()` in `dns-type.c`. Just this part of converting the RR types from string to ints as you suggestion is the last part.

poettering commented on src/resolve/resolved-gperf.gperf:
```diff
@@ -33,3 +33,4 @@ Resolve.ResolveUnicastSingleLabel, config_parse_bool,                    0,
 Resolve.DNSStubListenerExtra,      config_parse_dns_stub_listener_extra, 0,                   offsetof(Manager, dns_extra_stub_listeners)
 Resolve.CacheFromLocalhost,        config_parse_bool,                    0,                   offsetof(Manager, cache_from_localhost)
 Resolve.StaleRetentionSec,         config_parse_sec,                     0,                   offsetof(Manager, stale_retention_usec)
+Resolve.RefuseRecordTypes,         config_parse_string,                  0,                   offsetof(Manager, refuse_record_types)
```
please add your own config_parse_refuse_record_types() function, and parse the passed field via extract_first_word() + dns_type_from_string() piece-meal into a Set*.

bluca replied:
still not resolved

muhammadn replied:
@bluca I know it is not resolved as it's commented earlier on and pushing it is counter productive, unless you have some suggestions to add on to @poettering . So have some patience, i am trying to understand the code and i just started looking into systemd code in less than a week. When it's done, it's done.

bluca replied:
I mentioned it because the "good to merge" label was set, but it turns out it was a misclick: https://github.com/systemd/systemd/pull/34165#issuecomment-2320649919

muhammadn replied:
@poettering So i've added the value to `Set*` and how do i get this value in `resolved-dns-stub.c` ? It's not in some struct to store so i do not know how to get this in `resolved-dns-stub.c` from `conf-parser.c`

Is there some function called `set_` to get this?

`refused_record_types` is an int array in this case.
```
        _cleanup_set_free_free_ Set *refused_records;
        r = set_consume(refused_records, TAKE_PTR(refused_record_types));
        if (r < 0)
                return r;
```

muhammadn replied:
@bluca Oh. okay. Sorry for the misunderstanding!

muhammadn replied:
@poettering nevermind. forgot about the resolved manager. Now i can store it there and use it.

poettering commented on src/resolve/resolved-dns-stub.h:
```diff
@@ -25,6 +25,8 @@ struct DnsStubListenerExtra {
         union in_addr_union address;
         uint16_t port;
 
+        Set *refuse_record_types;
```
this should not be part of the extra stub, but there should be a global field for this, i.e. in Manager objectwhere you defined the char* field right now

muhammadn replied:
@poettering added this by accident when trying to figure out where to put `Set`. I'll remove this shortly.

bluca commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -949,6 +950,21 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
                 return;
         }
 
+        if (m->refuse_record_types) {
```
no need for the "if" block, STRV_FOREACH will be a no-op already if empty

muhammadn replied:
@bluca removed this code.

bluca commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -949,6 +950,21 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
+        if (m->refuse_record_types) {
+                STRV_FOREACH(i, strv_split(m->refuse_record_types, ",")) {
+                        int refuse_record_type;
+                        // delete_chars is just in case the user accidentally use qoutes like "AAAA" instead of AAAA in configuration
```
`/* */` for comments

muhammadn replied:
@bluca removed this code as well. Will do proper commenting next time. :)

bluca commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -949,6 +950,21 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
+                STRV_FOREACH(i, strv_split(m->refuse_record_types, ",")) {
+                        int refuse_record_type;
+                        // delete_chars is just in case the user accidentally use qoutes like "AAAA" instead of AAAA in configuration
+                        char* record_type_string = delete_chars(*i,"\"");
```
missing space after `*i,`

muhammadn replied:
using `extract_first_word` now. this block is removed.

bluca commented on src/resolve/resolved.conf.in:
```diff
@@ -35,3 +35,4 @@
 #ReadEtcHosts=yes
 #ResolveUnicastSingleLabel=no
 #StaleRetentionSec=0
+#RefuseRecordTypes=AAAA,MX
```
This must match the default, hence add the key without any value

muhammadn replied:
@bluca done

poettering commented:
oops, that i added the green label was an accident. wanted to add needs-rework. sorry for the confusion

muhammadn commented:
@poettering This is working now except that there is `dns-type.c` duplicate and i am unsure how to add it in meson.

muhammadn commented:
> @poettering This is working now except that there is `dns-type.c` duplicate and i am unsure how to add it in meson. I'll add documentation and tests later on. need to catch up on sleep.

I've fixed the `dns-type.c` issue and added some configuration to copy that file during building in `shared/meson.build`.

@poettering Looks like builds are ok. 

Note that in the root folder's `meson.build` includes `src/resolve` as i also need `resolved-manager.h` and that header has a lot of dependencies and adding it in `shared/meson.build` can be cumbersome (a large number of header files).

@bluca About documentation, i've grepped the `docs` folder in the repository to look for something related to `resolved` but it turns on nothing. Where should i put this information? is `resolved.conf.in` suffice?

i've checked `test-conf-parser.c` but i don't think it unit tests fits for my parser's scope as it's based is string parser code, same return values.

muhammadn commented:
@bluca i missed out `strv.h` from my older commits is no longer needed in `resolved-dns-stub.c` so i've removed it. So it's more cleaner from unnecessary header.

Thank you!

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -949,6 +949,12 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
         }
 
+        if (set_contains(m->refuse_record_types, INT_TO_PTR(dns_question_first_key(p->question)->type))) {
+                log_debug("Got request that is refused, refusing.");
```
please include an indication which rr type we are talking about here, i.e. `dns_type_to_string(dns_question_first_key(p->question)->type)`

poettering commented on src/shared/conf-parser.c:
```diff
@@ -45,6 +45,8 @@
 #include "time-util.h"
 #include "utf8.h"
+#include "dns-type.h"
+#include "resolved-manager.h"
```
as per coding style, please sort includes alphabetically

poettering replied:
also, please do *not* include resolved-specific code in our generic (shared/) code.

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                           "Specified string contains invalid ASCII characters, ignoring: %s", strna(escaped));
+                return 0;
+        }
```
both of these blocks are not relevant. please drop.

these checks make sense if you store strings as strings. but given you parse/validate them via dns_type_from_string() and then store only the results it doesn't matter if the strings where utf8 clean or not.

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
+                void *data,
+                void *userdata) {
+
+        char **s = ASSERT_PTR(data);
```
uh? you are not storing any strings here.

muhammadn replied:
char **s is used to store the string value from the configuration

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
+
+        r = free_and_strdup_warn(s, empty_to_null(rvalue));
+        if (r < 0)
+                return r;
```
this seems to be left-over?

muhammadn replied:
still needed

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
+        int refused_record_type;
+
+        /* sanitize the string to exclude quotes */
+        extract_first_word(&record_type_raw_string, &record_type_string, NULL, EXTRACT_UNQUOTE);
```
misses error checking.

also: please use extract_first_word() in a loop, always parsing the next word as you go, and each time adding the newly parsed RR type to the set.

muhammadn replied:
done

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
+
+        /* Get int values of DNS type for example "AAAA" string to get int value and store in Set */
+        Set *refused_records;
+        refused_records = set_new(NULL);
```
lacks oom checking.

also, please do not allocate the Set object ahead of time. instead use set_ensure_put() which allocates the set when the first item is added.

in our tree we generally try to optimize empty `Set` objects (and other collection types) away, in the sense that NULL pointers are considered equivalent to empty collections, in most ways. Thus, we typically just assign A Set* to NULL initially, and then fill it up via set_ensure_put() which actually allocates the Set* only when it is supposed to contain at least one entry.

muhammadn replied:
done

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
+        Set *refused_records;
+        refused_records = set_new(NULL);
+
+        STRV_FOREACH(i, strv_split(record_type_string, ",")) {
```
strv_split() is internally just a loop around extract_first_word(), turning it into a char** string array. please don't do that. instead just do your own loop around extract_first_word() and parse each dns type returned as you go without keeping the allocated strings around

also you are leaking the result of the strv_split() here

muhammadn replied:
not using strv_split anymore, using own loop

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
+        refused_records = set_new(NULL);
+
+        STRV_FOREACH(i, strv_split(record_type_string, ",")) {
+                refused_record_type = dns_type_from_string(*i);
```
this might very well fail (because the user specified an unknown rr type). please log about it (via log-syntax), but ignore it otherwise (i.e. do not consider this fatal), for compatibility.

muhammadn replied:
done

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
+
+        STRV_FOREACH(i, strv_split(record_type_string, ",")) {
+                refused_record_type = dns_type_from_string(*i);
+                set_put(refused_records, INT_TO_PTR(refused_record_type));
```
oom check missing.

muhammadn replied:
won't need this as everything moved to resolve/resolved-resolve-conf.c, added oom check

poettering commented on src/shared/meson.build:
```diff
@@ -311,6 +311,50 @@ ethtool_link_mode_xml = custom_target(
+        output : fname,
+        command : [cp, '@INPUT0@', '@OUTPUT@'],
+        capture : false)
+shared_sources += dns_type_c
```
this all should be unnecessary if you place config_parse_refuse_record_types() in src/rsolve/resolved-conf.c

muhammadn replied:
won't need this as everything moved to resolve/

poettering commented on src/shared/conf-parser.c:
```diff
@@ -2119,3 +2121,72 @@ int config_parse_ip_protocol(
         return 1; /* done. */
 }
+
+int config_parse_refuse_record_types(
```
(src/shared/conf-parser.conf is the place for super generic parsers only, not for specific ones. Typically this means that parsers placed here need to have at least to users in our tree, in different components).

poettering commented:
please squash your commits.

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -949,6 +949,12 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
+                log_debug("Got request that is refused, refusing.");
+                dns_stub_send_failure(m, l, s, p, DNS_RCODE_REFUSED, false);
+                return;
+        }
```
so this will only refuse these look ups when done via the DNS stubs. But we also provide lookups via a dbus ipc api and via varlink ipc api. These need to refuse the requests as well.

I think it would be best to move the check into the top of dns_query_new(), and return a recognizable error there, which is then turned into DNS_RCODE_REFUSED in the stub, and into appropriate dbus/varlink errors in those ipc interfaces.

muhammadn replied:
@poettering Most of your recommendations above is implemented but i don't understand the last part about dbus/varlink errors.

muhammadn commented:
> please squash your commits.

I'll squash but i need to know that everything is ok before i do this

yuwata commented on meson.build:
```diff
@@ -702,6 +702,7 @@ env = find_program('env')
 rsync = find_program('rsync', required : false)
 diff = find_program('diff')
 find = find_program('find')
+cp = find_program('cp')
```
Unused? Please drop this.

muhammadn replied:
Yes. unused, some leftover code i forgot to remove.

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+                void *data,
+                void *userdata) {
+
+        Manager *m = ASSERT_PTR(userdata);
+
+        int r;
```
`data` points to the `refuse_record_types`. So,
```c
Set **types = ASSERT_PTR(data);
```

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+        int refused_record_type;
+
+        /* Get int values of DNS type for example "AAAA" string to get int value and store in Set */
+        Set *refused_records = NULL;
```
This is leaked on failure. Please use `_cleanup_set_free_` attribute.
BTW, I do not think this is necessary. See comment below.

muhammadn replied:
`_cleanup_set_free_` will cause a segfault where the data was freed before it can be used by `Manager`

poettering replied:
once you finished putting together the `Set*` object with all its contents, you should install it on the manager, and set it to NULL here, so that the `_cleanup_` handler is "deactivated", if you so will. This is typically done via a call to the `TAKE_PTR()` macro. That macro just shortens this: `p = q; q = NULL;` to `p = TAKE_PTR(q);` but does the same thing, i.e. assignment + invalidation of the old pointer

muhammadn replied:
Trying to understand this part. I'll think about it first.

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+        if (r < 0)
+               return r;
+
+        char* record_type_string;
```
We usually use `char *foo`, rather than `char* foo`.

muhammadn replied:
@yuwata code no longer necessary, removed this as well.

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+                return 1;
+        }
+
+        r = free_and_strdup_warn(data, empty_to_null(rvalue));
```
What? Please drop. This leaks the copied string.

muhammadn replied:
Drop everything related to `data`  and use `rvalue` instead so this code is no longer necessary.

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+        assert(rvalue);
+
+        if (isempty(rvalue)) {
+                data = mfree(data);
```
Then,
```c
*types = set_free(*types);
```

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+        /* Get int values of DNS type for example "AAAA" string to get int value and store in Set */
+        Set *refused_records = NULL;
+
+        for (;;) {
+               r = extract_first_word(data, &record_type_string, ",", 0);
+               if (r < 0)
+                      return r;
```
Please see other conf parsers e.g. in shared/conf-parser.c that uses `extract_first_word()`.
```c
for (const char *p = rvalue;;) {
        _cleanup_free_ char *word = NULL;

        r = extract_first_word(&p, &word, ",", 0);
        if (r < 0)
                return log_syntax_parse_error(unit, filename, line, r, lvalue, rvalue);
        if (r == 0)
                break;

        ...
}
```

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+               if (r == 0)
+                       break;
+
+               refused_record_type = dns_type_from_string(record_type_string);
+               /* log error of invalid dns type */
+               if (refused_record_type == _DNS_TYPE_INVALID)
+               log_error("dns type is invalid");
```
Please do not store the invalid type, and use log_syntax():
```c
int t = dns_type_from_string(word);
if (t < 0) {
        log_syntax(unit, LOG_WARNING, filename, line, t, "Invalid DNS record type, ignoring: %s", word);
        continue;
}
```

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+               log_error("dns type is invalid");
+
+               r = set_ensure_put(&refused_records, NULL, INT_TO_PTR(refused_record_type));
+               if (r == -ENOMEM)
+                      return log_oom();
```
Please do not ignore other error code (though, it should only return -ENOMEM on error):
```c
if (r < 0)
        return log_oom();
```

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,61 @@ int manager_parse_config_file(Manager *m) {
+                      return log_oom();
+        }
+
+        m->refuse_record_types = refused_records;
```
Previous assignment is ignored, and leaked. I think it is OK to append the previous assignment, So, in the above loop,
```c
r = set_ensure_put(types, NULL, INT_TO_PTR(t));
```

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -480,6 +480,14 @@ int dns_query_new(
         assert(m);
 
+        /* Check for records that is refused and refuse query for the records if matched in configuration */
+        DNS_QUESTION_FOREACH(key, question_utf8) {
```
Please drop unnecessary `{}`.

muhammadn replied:
@yuwata Yeah, misread this code, it was one block so yeah, i've removed the curly braces.

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -480,6 +480,14 @@ int dns_query_new(
+        /* Check for records that is refused and refuse query for the records if matched in configuration */
+        DNS_QUESTION_FOREACH(key, question_utf8) {
+                if (set_contains(m->refuse_record_types, INT_TO_PTR(key->type))) {
+                        log_debug("Got request for %s record that is refused.", dns_type_to_string(key->type));
+                        return -ENOSYS;
```
```c
return log_debug_errno(SYNTHETIC_ERRNO(ENOSYS), ...);
```

But, I do not think ENOSYS is not a good error code here. Please use more recognizable error code, e.g. `ENOANO` or so.

muhammadn replied:
@yuwata used ENOANO now

yuwata commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -992,6 +992,13 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
+
+        /* Refuse query if there is -ENOSYS */
+        if (r == -ENOSYS) {
+                dns_stub_send_failure(m, l, s, p, DNS_RCODE_REFUSED, false);
+                return;
```
return (void) dns_stub_send_failure(...);

muhammadn replied:
Did as recommended.

yuwata commented on src/resolve/resolved-manager.h:
```diff
@@ -137,6 +137,9 @@ struct Manager {
         bool read_etc_hosts;
 
+        /* List of refused DNS Record Types*/
+        Set *refuse_record_types;
```
This is not freed in `manager_free()`.

muhammadn replied:
@yuwata I thought this is done by `resolved-manager.c` in `manager_new()` with:

```
        _cleanup_(manager_freep) Manager *m = NULL;
````

poettering replied:
You have to free the set explicitly in manager_free(), so that it is freed whenever the manager is freed.

muhammadn replied:
I'm not sure about how to do this. Just that freeing in `src/resolve/resolved-conf.c` will cause the data in the manager to be lost and cause segfaults.

yuwata replied:
Please add a single line `set_free(m->refuse_record_types);` in `manager_free()`.

yuwata commented on src/shared/conf-parser.h:
```diff
@@ -278,6 +278,7 @@ CONFIG_PARSER_PROTOTYPE(config_parse_uint32_invert_flag);
 CONFIG_PARSER_PROTOTYPE(config_parse_id128);
 CONFIG_PARSER_PROTOTYPE(config_parse_tristate);
 CONFIG_PARSER_PROTOTYPE(config_parse_string);
+CONFIG_PARSER_PROTOTYPE(config_parse_refuse_record_types);
```
Please move this to resolved-conf.h

muhammadn replied:
@yuwata This was some leftover code which i forgot to move it over.  this is moved.

yuwata reviewed (changes requested):
Several superficial comments.

poettering commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,46 @@ int manager_parse_config_file(Manager *m) {
+                     continue;
+               }
+
+               r = set_ensure_put(&refused_records, NULL, INT_TO_PTR(r));
```
honestly, just add what you parsed here directly to `m->refuse_record_types`.

I mean, if you do:

```
RefuseRecordTypes=AAAA MX
RefuseRecordTypes=A
```

then this should be equivalent to:

```
RefuseRecordTypes=A AAAA MX
```

And in your current code it would be, the 2nd line would completely override the first, but is not merged like it should (and how this is usually done in our config files)

muhammadn replied:
done.

poettering commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,46 @@ int manager_parse_config_file(Manager *m) {
+        Set *refused_records = NULL;
+
+        refused_records = set_free(refused_records);
+
```
in our codebase, if you assign an empty string to a setting it should generally reset the field to its default. Hence, please add:

```c
if (isempty(rvalue)) {
       m->refuse_record_types = set_free(m->refuse_record_types);
       return 0;
```

This is nice because it allows that configuration file drop-ins can override each other and the main config file.

muhammadn replied:
Yeah. I've updated the code to reflect on it. Thank you!

poettering commented on src/resolve/resolved-dns-query.c:
```diff
@@ -480,6 +480,12 @@ int dns_query_new(
+        DNS_QUESTION_FOREACH(key, question_utf8)
+                if (set_contains(m->refuse_record_types, INT_TO_PTR(key->type))) {
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOANO), "Got request for %s record that is refused.", dns_type_to_string(key->type));
+                }
```
coding style: no `{}` around single ine if blocks.

Hmm, so this is not going to work. The only reason why DnsQuestion structures contain multiple DnsResourceKey objects is to permit combined A + AAAA queries (because clients typically want to query both at once). However, if such a question is issued this loop of yours here, will refuse the whole query, even though instead we should simply suppress the relevant record type from the question, and if the result is non-empty continue.

Hence, maybe introduce a new helper call here, that "mangles" a DnsQuestion object: if any of the rr types is listed among the rr types to refuse it will remove them from the question, unless this would make the question empty in which case it would return an error.

Note that DnsResourceKey and DnsQuestion are supposed to be immutable once constructed fully. This means if you realize you need to remove a key from the question, then you need to allocate a new question object and initialize it from what's left.

muhammadn replied:
I'll look into this but is it just me or nslookup sends two question for every record type?

![image](https://github.com/user-attachments/assets/32ab1e48-8424-4485-b4fb-b5c86037ec7c)

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -992,6 +992,12 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
+        /* Refuse query if there is -ENOSYS */
+        if (r == -ENOANO) {
+                return (void) dns_stub_send_failure(m, l, s, p, DNS_RCODE_REFUSED, false);
+        }
```
No {} around single line if blocks

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -992,6 +992,12 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
                                   SD_RESOLVED_CLAMP_TTL);
+
+        /* Refuse query if there is -ENOSYS */
+        if (r == -ENOANO) {
```
if check and comment ar out of sync

poettering commented:
also needs a man page update, explaining the new option

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -992,6 +992,12 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
+        if (r == -ENOANO) {
+                return (void) dns_stub_send_failure(m, l, s, p, DNS_RCODE_REFUSED, false);
+        }
+
```
similar handling for ENOANO should be added for dns_query_new() calls in src/resolve/resolved-bus.c and src/resolve/resolved-varlink.c, returning a recognizable error.

in the former add a local wrapper dns_query_new_for_bus() or so, that fills in a useful sd_bus_error structure in case the ENOANO thing is seen. Then switch all callers over from dns_query_new() to this new wrapper call. 

Similar for the varlink case.

(background: we answer lookup requests via three mechanisms: D-Bus IPC, Varlink IPC and the local stub). Right now the latter would return a clean REFUSED error, but the others should return clean, descriptive errors for this too.

muhammadn commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,46 @@ int manager_parse_config_file(Manager *m) {
+               r = extract_first_word(&p, &word, ",", EXTRACT_UNQUOTE);
+               if (r < 0)
+                       return log_syntax_parse_error(unit, filename, line, r, lvalue, rvalue);
+
```
Thanks for the pointers. I'll take a note of it. 👍

muhammadn commented on src/resolve/resolved-conf.c:
```diff
@@ -414,3 +414,46 @@ int manager_parse_config_file(Manager *m) {
+
+        for (const char *p = rvalue;;) {
+               _cleanup_free_ char *word = NULL;
+               r = extract_first_word(&p, &word, ",", EXTRACT_UNQUOTE);
```
I think i'll just use whitespaces since you agree on that.

muhammadn commented:
> also needs a man page update, explaining the new option

added information in the man page.

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -413,3 +413,47 @@ int manager_parse_config_file(Manager *m) {
+
+        if (isempty(rvalue)) {
+                m->refuse_record_types = set_free(m->refuse_record_types);
+                return 0;
```
Not a matter, but for new conf parsers, please return 1 on success.
`return 0` means non-critical error in most cases.

muhammadn replied:
Yes. Just that i thought normal C would return 0 for success.

yuwata replied:
`return 1` please.

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -413,3 +413,47 @@ int manager_parse_config_file(Manager *m) {
+                      return log_oom();
+        }
+
+        return 1;
```
As said. Please drop this.

yuwata replied:
Please drop.

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -413,3 +413,47 @@ int manager_parse_config_file(Manager *m) {
         return 0;
 }
+
+int config_parse_refuse_record_types(
```
Not a matter, but this parser itself is generic, and not specific for `refuse`.
Please rename `config_parse_record_types()`.

yuwata commented on src/resolve/resolved-conf.c:
```diff
@@ -413,3 +413,47 @@ int manager_parse_config_file(Manager *m) {
+                void *data,
+                void *userdata) {
+
+        Manager *m = ASSERT_PTR(userdata);
```
Then, please take a pointer to `Set` from `data`, rather than `Manager` from `userdata`:
```c
Set *types = ASSERT_PTR(data);
```

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -524,6 +524,11 @@ int dns_query_new(
         assert(m);
 
+        /* Check for records that is refused and refuse query for the records if matched in configuration */
+        DNS_QUESTION_FOREACH(key, question_utf8)
```
Is it not necessary to check `question_idna`?

yuwata commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -996,6 +996,11 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
                                   (DNS_PACKET_CD(p) ? SD_RESOLVED_NO_VALIDATE | SD_RESOLVED_NO_CACHE : 0)|
                                   (DNS_PACKET_DO(p) ? SD_RESOLVED_REQUIRE_PRIMARY : 0)|
                                   SD_RESOLVED_CLAMP_TTL);
+
```
Please drop the empty line.

yuwata commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -996,6 +996,11 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
                                   (DNS_PACKET_DO(p) ? SD_RESOLVED_REQUIRE_PRIMARY : 0)|
                                   SD_RESOLVED_CLAMP_TTL);
+
+        /* Refuse query if there is -ENOANO */
+        if (r == -ENOANO)
```
Please move the comment in the `if` line:
```c
if (r == -ENOANO) /* Refuse query if there is -ENOANO.*/
        return (void) dns_stub_send_failure(m, l, s, p, DNS_RCODE_REFUSED, false);
if (r < 0) {
```

yuwata commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -996,6 +996,11 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
+        /* Refuse query if there is -ENOANO */
+        if (r == -ENOANO)
+                return (void) dns_stub_send_failure(m, l, s, p, DNS_RCODE_REFUSED, false);
+
```
Please drop the empty line.

yuwata commented on src/resolve/resolved-varlink.c:
```diff
@@ -37,6 +37,23 @@ static void lookup_parameters_destroy(LookupParameters *p) {
+
+        int r = dns_query_new(m, ret, question_utf8, question_idna, question_bypass, ifindex, flags);
+        if (r == -ENOANO) {
+                sd_varlink_error((*ret)->varlink_request, "io.systemd.Resolve.DNSError", NULL);
```
No no, `*ret` is `NULL` at this point.
As similar to the DBus case, please add one more argument `sd_varlink *link`, then use it.
Also, propagate error from `sd_varlink_error()`.
```c
return sd_varlink_error(link, "io.systemd.Resolve.DNSError", NULL);
```

yuwata reviewed (changes requested):
Please add test cases in `TEST-75-RESOLVED`.

muhammadn commented:
> Please add test cases in `TEST-75-RESOLVED`.

Thanks for being patient with me.

I do not understand what it means with `TEST-75-RESOLVED` as i see `test/TEST75-RESOLVED` only has meson.build file that reads some other files.

Where should i add test cases?

Thank you.

muhammadn commented:
@yuwata for test cases, is it in `test/units/TEST-75-RESOLVED.sh`?

Just to confirm because the test cases isn't that much for me to make sense of.

poettering commented on man/resolved.conf.xml:
```diff
@@ -317,6 +317,17 @@
+        <programlisting>RefuseRecordTypes=AAAA SRV TXT</programlisting>
+        </para>
+
+        <xi:include href="version-info.xml" xpointer="v257"/></listitem>
```
update to v258

poettering commented on src/resolve/resolved-conf.c:
```diff
@@ -413,3 +413,45 @@ int manager_parse_config_file(Manager *m) {
+               if (r < 0)
+                     return log_syntax_parse_error(unit, filename, line, r, lvalue, rvalue);
+               if (r == 0)
+                     return 1;
```
our indentation is 8ch, this appears to be 6ch?

muhammadn replied:
Yeah, the `if` function line also lacked one more space. I've fixed all (hopefully)

yuwata commented:
> @yuwata for test cases, is it in `test/units/TEST-75-RESOLVED.sh`?
> 
> Just to confirm because the test cases isn't that much for me to make sense of.

Right.

poettering commented on src/resolve/resolved-dns-query.c:
```diff
@@ -524,6 +524,15 @@ int dns_query_new(
+
+        DNS_QUESTION_FOREACH(key, question_idna)
+                if (set_contains(m->refuse_record_types, INT_TO_PTR(key->type)))
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOANO), "Got request for %s record that is refused.", dns_type_to_string(key->type));
```
please split out the check into a helper func, i.e.

```c
bool dns_question_has_record_types(DnsQuestion *q, Set *types) { 
        …
}
```

so that these two loops here become:

```c
if (dns_question_has_record_types(question_utf8, m->refuse_record_types) ||
        dns_question_has_record_types(question_idna, m->refuse_record_types))
        return log_debug_errno(…
```

you get the idea

poettering replied:
hmm, so thining about this: i think this needs some tweaking. DnsQuestion can contain multiple lookup keys primarily to cover for cases where we want to look for A and AAAA for some domain at the same time. Given that this is the primary usecase I think we really should tweak behaviour here so that:

1. when *all* of the request types in the DnsQuestion are among the record types to refuse we return a failure, as your current code does it
2. when however only *some* of the request types in the DnsQuestion are to be refused we should just silently remove them from the question, and continue with the ones that are allowed.

muhammadn replied:
Need sometime to digest this. I'll try to reread it again to better understand it.

muhammadn replied:
Realised this when i tried to query using resolvectl when trying to build tests for resolvectl. So maybe this is your concern, i'll take a look on how to go about this.

```
$ resolvectl query google.com
google.com: resolve call failed: DNS query type refused.
```

poettering commented on src/resolve/resolved-dns-query.c:
```diff
@@ -1377,3 +1382,12 @@ bool dns_query_fully_authoritative(DnsQuery *q) {
+        DnsResourceKey *key;
+
+        DNS_QUESTION_FOREACH(key, q)
+                return (set_contains(types, INT_TO_PTR(key->type)));
```
the outer () are redundant.

with this you are now only checking the first entry...

poettering commented on src/resolve/resolved-gperf.gperf:
```diff
@@ -34,3 +34,4 @@ Resolve.ResolveUnicastSingleLabel, config_parse_bool,                    0,
 Resolve.DNSStubListenerExtra,      config_parse_dns_stub_listener_extra, 0,                   offsetof(Manager, dns_extra_stub_listeners)
 Resolve.CacheFromLocalhost,        config_parse_bool,                    0,                   offsetof(Manager, cache_from_localhost)
 Resolve.StaleRetentionSec,         config_parse_sec,                     0,                   offsetof(Manager, stale_retention_usec)
+Resolve.RefuseRecordTypes,         config_parse_record_types,     0,                   offsetof(Manager, refuse_record_types)
```
please align columns

muhammadn replied:
aligned.

poettering commented on src/resolve/resolved-varlink.c:
```diff
@@ -37,6 +37,22 @@ static void lookup_parameters_destroy(LookupParameters *p) {
+                int ifindex,
+                uint64_t flags) {
+
+        int r = dns_query_new(m, ret, question_utf8, question_idna, question_bypass, ifindex, flags);
```
`int r;` is kinda special to us, we always declare it in a separate line as first thing in functions where it is needed. Only `r` gets that treatment, because it's like *the* variable we return function return values/error codes in.

muhammadn replied:
ah. ok, i'll look into this.

muhammadn commented:
> > @yuwata for test cases, is it in `test/units/TEST-75-RESOLVED.sh`?
> > Just to confirm because the test cases isn't that much for me to make sense of.
> 
> Right.

Done. Added new test case.

yuwata commented:
CIs are unhappy
```
871/1948 systemd:resolve / test-dns-query                                                  FAIL             0.09s   killed by signal 6 SIGABRT
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
/* test_dns_query_new_single_question */
Assertion 'q' failed at src/resolve/resolved-dns-query.c:1387, function dns_question_has_record_types(). Aborting.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```

yuwata commented on man/resolved.conf.xml:
```diff
@@ -317,6 +317,17 @@
 
+      <varlistentry>
+        <term><varname>RefuseRecordTypes=</varname></term>
+        <listitem><para>Takes a string argument separated by spaces. This option is to refuse (deny) record types
+         set from from being queried</para>
```
Please write something like the following:
`Takes a list of DNS record types separated by space. Specified record types in each query will be refused. This option can be specified multiple times. If an empty string is specified, then all previous assignments are cleared.`

muhammadn replied:
@yuwata done.

yuwata commented on src/resolve/resolved-bus.c:
```diff
@@ -29,6 +29,21 @@
+                DnsPacket *question_bypass,
+                int ifindex,
+                uint64_t flags,
+                sd_bus_error *error) {
+        int r = dns_query_new(m, ret, question_utf8, question_idna, question_bypass, ifindex, flags);
```
Please insert an empty line after the function arguments, and declare `r` separately.
```c
                sd_bus_error *error) {

        int r;

        r = dns_query_new(...);
        if (r == -ENOANO)
```

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -1377,3 +1382,12 @@ bool dns_query_fully_authoritative(DnsQuery *q) {
 }
+
+bool dns_question_has_record_types(DnsQuestion *q, Set *types) {
+        assert(q);
```
This assertion is not necessary. `DNS_QUESTION_FOREACH()` handles `NULL` gracefully.

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -1377,3 +1382,12 @@ bool dns_query_fully_authoritative(DnsQuery *q) {
+bool dns_question_has_record_types(DnsQuestion *q, Set *types) {
+        assert(q);
+        DnsResourceKey *key;
+
```
Please drop the empty line.

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -1377,3 +1382,12 @@ bool dns_query_fully_authoritative(DnsQuery *q) {
+        DnsResourceKey *key;
+
+        DNS_QUESTION_FOREACH(key, q)
+                return set_contains(types, INT_TO_PTR(key->type));
```
This checks only the first key. Return only when `set_contains()` is true.
```c
DnsResourceKey *key;
DNS_QUESTION_FOREACH(key, q)
        if (set_contains(types, INT_TO_PTR(key->type)))
                return true;

return false;
```

yuwata commented on src/resolve/resolved-varlink.c:
```diff
@@ -37,6 +37,22 @@ static void lookup_parameters_destroy(LookupParameters *p) {
+                DnsQuestion *question_utf8,
+                DnsQuestion *question_idna,
+                DnsPacket *question_bypass,
+                sd_varlink *link,
+                int ifindex,
+                uint64_t flags) {
```
Please make the ordering consistent with `dns_query_new_for_dbus()`, that is, move `sd_varlink *link` to the last.

yuwata commented on src/resolve/resolved-varlink.c:
```diff
@@ -37,6 +37,22 @@ static void lookup_parameters_destroy(LookupParameters *p) {
+                int ifindex,
+                uint64_t flags) {
+
+        int q = dns_query_new(m, ret, question_utf8, question_idna, question_bypass, ifindex, flags);
```
Please use `int r`, rather than `q`.
In our code base `int r` is a kind of special variable. Please use it.

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -1377,3 +1382,12 @@ bool dns_query_fully_authoritative(DnsQuery *q) {
         return (q->answer_query_flags & SD_RESOLVED_FROM_MASK & ~(SD_RESOLVED_FROM_TRUST_ANCHOR | SD_RESOLVED_FROM_ZONE)) == 0;
 }
+
+bool dns_question_has_record_types(DnsQuestion *q, Set *types) {
```
I do not think this should be exported. Please make it static and move just before the declaration of `dns_query_new()`.

muhammadn replied:
removed export

yuwata commented on test/units/TEST-75-RESOLVED.sh:
```diff
@@ -1084,6 +1084,29 @@ testcase_13_varlink_subscribe_dns_configuration() {
 }
 
+# Test RefuseRecordTypes
+testcase_14_refuse_record_types() {
```
Please add cleanup function something like the following:
```sh
    cleanup() {
        rm -f /run/systemd/resolved.conf.d/refuserecords.conf
        if [[ -e /etc/resolv.conf.bak ]]; then
            rm -f /etc/resolv.conf
            mv /etc/resolv.conf.bak /etc/resolv.conf
        fi
        restart_resolved
    }
    trap cleanup RETURN ERR
```
Otherwise, the setting may be used other test cases.

yuwata commented on test/units/TEST-75-RESOLVED.sh:
```diff
@@ -1084,6 +1084,29 @@ testcase_13_varlink_subscribe_dns_configuration() {
+        echo "[Resolve]"
+        echo "RefuseRecordTypes=AAAA SRV TXT"
+    } >/run/systemd/resolved.conf.d/refuserecords.conf
+    ln -svf /run/systemd/resolve/stub-resolv.conf /etc/resolv.conf
```
Please save the existing `/etc/resolv.conf`, and restore it on cleanup:
```sh
if [[ -e /etc/resolv.conf ]]; then
    mv /etc/resolv.conf /etc/resolv.conf.bak
fi
ln -s /run/systemd/resolve/stub-resolv.conf /etc/resolv.conf
```

muhammadn replied:
done

yuwata commented on test/units/TEST-75-RESOLVED.sh:
```diff
@@ -1084,6 +1084,29 @@ testcase_13_varlink_subscribe_dns_configuration() {
+    grep -qF "status: REFUSED" "$RUN_OUT"
+
+    run dig localhost -t A
+    grep -qF "status: NOERROR" "$RUN_OUT"
```
Please add test cases with `resolvectl query`

muhammadn replied:
done.

yuwata reviewed (changes requested):
Looks much much better now. But one more round please.

muhammadn commented:
> CIs are unhappy
> 
> ```
> 871/1948 systemd:resolve / test-dns-query                                                  FAIL             0.09s   killed by signal 6 SIGABRT
> ――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
> stderr:
> /* test_dns_query_new_single_question */
> Assertion 'q' failed at src/resolve/resolved-dns-query.c:1387, function dns_question_has_record_types(). Aborting.
> ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
> ```

Thanks for digging into the build logs for me.

i was looking for the logs but realised the log artifacts in summary. will check that from now.  👍

yuwata commented on test/units/TEST-75-RESOLVED.sh:
```diff
@@ -1084,6 +1084,42 @@ testcase_13_varlink_subscribe_dns_configuration() {
 
+# Test RefuseRecordTypes
+testcase_14_refuse_record_types() {
+    cleanup() {
```
Please add the comment to suppress shellcheck warning. See `testcase_13_varlink_subscribe_dns_configuration()` in the above for example.
```
# shellcheck disable=SC2317
```

poettering commented:
The newest version still will refuse the entire query even if one of the RR types in the DnsQuestion would actually be allowed. 

hence, please change the code to refuse the question entirely only if *all* RR types are among the refusal list. And if only some RR types are such listed mangle the DnsQuestion (i.e. create a copy, and copy over what is allowed), and proceed with that.

As mentioned, it's a pretty typical case to query both A and AAAA at the same time, that's after all what ResolveHostname() varlink call does. And if the refusal says "no aaaa" then we should still use "a".

muhammadn commented:
@poettering Not that i've tried. Spent the entire night trying to do what you wanted.

```
static bool dns_question_has_record_types(DnsQuestion *q, Set *types) {
        DnsResourceKey *k;
        _cleanup_set_free_ Set *records = NULL, *allow = NULL, *t = NULL, *s = NULL;

        /* convert DnsQuestion to Set */
        DNS_QUESTION_FOREACH(k, q)
                        set_ensure_put(&records, NULL, INT_TO_PTR(k));

        /* this does not make sense */
        SET_FOREACH(t, types)
                SET_FOREACH(s, records)
                     if (s == t)
                        set_ensure_put(&allow, NULL, INT_TO_PTR(t));  

        return false;
}
```

I've looked at `set_equal()` but that's comparing sets which are _ordered_. If it allows unordered sets, i think i can just do `set_equal(t, s)` but i don't think that's possible for now.

poettering commented:
hmm? just iterate through the DnsQuestion, and maintain an enum: 

```c
static manager_validate_and_mangle_question(Manager *m, DnsQuestion **question) {
        enum {
                VALID_DONT_KNOW,
                VALID_ALL_GOOD, 
                VALID_ALL_BAD,
                VALID_MIXED,
        } good = VALID_DONT_KNOW;

        assert(m);
        assert(question); 
                
        /* Check if the query should be refused entirely, accepted entirely, or needs to be mangled to suppress RR types not allowed */
        DNS_QUESTION_FOREACH(key, *question) {
                 if (set_contains(m->refuse_record_types, INT_TO_PTR(key->type))) 
                         good = IN_SET(good, VALID_DONT_KNOW, VALID_ALL_BAD) ? VALID_ALL_BAD : VALID_MIXED;
                 else 
                         good = IN_SET(good, VALID_DONT_KNOW, VALID_ALL_GOOD) ? VALID_ALL_GOOD : VALID_MIXED;
        }

        if (good == VALID_ALL_GOOD) 
                return 0; /* All good, no need to mangle */
        if (good != VALID_MIXED)
                return -ENOANO; /* This is fully bad or empty? Refuse */

        /* Mangle the question suppressing bad entries, leaving good entries */
        _cleanup_(dns_question_unrefp) DnsQuestion *new_question = dns_question_new(dns_question_size(*question));
        if (!new_question)
                return -ENOMEM;

        DnsQuestionItem *item;  
        DNS_QUESTION_FOREACH_ITEM(item, *question) {
                 if (set_contains(m->refuse_record_types, INT_TO_PTR(item->key->type))) 
                         continue;
                 r = dns_question_add_raw(new_question, item->key, item->flags);
                 if (r < 0)
                         return r;
        }

        dns_question_unref(*question);
        *question = TAKE_PTR(new_question);
}
```

entirely untested...

muhammadn commented:
> hmm? just iterate through the DnsQuestion, and maintain an enum:
> 
> ```c
> static manager_validate_and_mangle_question(Manager *m, DnsQuestion **question) {
>         enum {
>                 VALID_DONT_KNOW,
>                 VALID_ALL_GOOD, 
>                 VALID_ALL_BAD,
>                 VALID_MIXED,
>         } good = VALID_DONT_KNOW;
> 
>         assert(m);
>         assert(question); 
>                 
>         /* Check if the query should be refused entirely, accepted entirely, or needs to be mangled to suppress RR types not allowed */
>         DNS_QUESTION_FOREACH(key, *question) {
>                  if (set_contains(m->refuse_record_types, INT_TO_PTR(key->type))) 
>                          good = IN_SET(good, VALID_DONT_KNOW, VALID_ALL_BAD) ? VALID_ALL_BAD : VALID_MIXED;
>                  else 
>                          good = IN_SET(good, VALID_DONT_KNOW, VALID_ALL_GOOD) ? VALID_ALL_GOOD : VALID_MIXED;
>         }
> 
>         if (good == VALID_ALL_GOOD) 
>                 return 0; /* All good, no need to mangle */
>         if (good != VALID_MIXED)
>                 return -ENOANO; /* This is fully bad or empty? Refuse */
> 
>         /* Mangle the question suppressing bad entries, leaving good entries */
>         _cleanup_(dns_question_unrefp) DnsQuestion *new_question = dns_question_new(dns_question_size(*question));
>         if (!new_question)
>                 return -ENOMEM;
> 
>         DnsQuestionItem *item;  
>         DNS_QUESTION_FOREACH_ITEM(item, *question) {
>                  if (set_contains(m->refuse_record_types, INT_TO_PTR(item->key->type))) 
>                          continue;
>                  r = dns_question_add_raw(new_question, item->key, item->flags);
>                  if (r < 0)
>                          return r;
>         }
> 
>         dns_question_unref(*question);
>         *question = TAKE_PTR(new_question);
> }
> ```
> 
> entirely untested...

Thanks! I've update the code a bit. Didn't know i can use `IN_SET` what i did was to look into `src/tests` to understand how the helper functions work. 

I think i did saw systemd documention on hacking the code, i think i'll need to read more on that as well. I am really sorry for troubling you.

This work was initially because i'm working for a financial company and AAAA requests can frustrate our network engineers with the spurious AAAA requests. Dropped this work midway because i was burned out and now i have some spare time to work on this.

yuwata commented:
@muhammadn Making this draft tentatively. Please push 'ready' button when you request a review.

muhammadn commented:
@yuwata  👌 Ok thank you!, I'll working on the tests now. 👍

muhammadn commented:
@yuwata `mkiso` build is failing. Not sure what is the issue. Also the issue with `ABSAN/UBSAN` unit test compilation modes.

Other than that, other unit tests have passed.

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -524,6 +573,11 @@ int dns_query_new(
         assert(m);
 
+        /* Check for records that is refused and refuse query for the records if matched in configuration */
+        if (dns_question_has_record_types(question_utf8, m->refuse_record_types) ||
```
Please directly call `manager_validate_and_mangle_question()` and propagate the error code:
```c
r = validate_and_mangle_question(question_utf8, m->refuse_record_types);
if (r < 0)
        return r;

r = validate_and_mangle_question(question_idna, m->refuse_record_types);
if (r < 0)
        return r;
```

yuwata commented on src/resolve/resolved-dns-query.c:
```diff
@@ -508,6 +508,55 @@ DnsQuery *dns_query_free(DnsQuery *q) {
 }
 
+/* Almost all of this is not my code, it's help from @poettering */
+static int manager_validate_and_mangle_question(Set *types, DnsQuestion **question) {
```
The main target argument is `DnsQuestion`, and this does not take `Manager` object.
Hence, please rename and reorder argument:
```c
static int validate_and_mangle_question(DnsQuestion **question, Set *types) {
```

yuwata commented:
@muhammadn Made this PR draft again. Please push the 'ready for review' button again when the suggestions are applied.

muhammadn commented:
@yuwata Thank you! It's ready for review, the `mkosi arch` error is due to mount feature and i think it's not related? For the rawhide error is due to hostname checks.

yuwata commented on src/resolve/resolved-bus.c:
```diff
@@ -1208,7 +1226,7 @@ static int resolve_service_hostname(DnsQuery *q, DnsResourceRecord *rr, int ifin
-        r = dns_query_new(q->manager, &aux, question, question, NULL, ifindex, q->flags|SD_RESOLVED_NO_SEARCH);
+        r = dns_query_new_for_bus(q->manager, &aux, question, question, NULL, ifindex, q->flags|SD_RESOLVED_NO_SEARCH, NULL);
         if (r < 0)
                 return r;
```
Setting `NULL` for `error` in `dns_query_new_for_bus()` is meaningless. Please use `dns_query_new()` and `reply_method_errorf()` on failure.
```c
r = dns_query_new(...);
if (r == -ENOANO)
        return reply_method_errorf(q, BUS_ERROR_DNS_REFUSED, "DNS query type refused.");
if (r < 0)
        return r;
```

yuwata commented on src/resolve/resolved-varlink.c:
```diff
@@ -37,6 +37,24 @@ static void lookup_parameters_destroy(LookupParameters *p) {
+
+        r = dns_query_new(m, ret, question_utf8, question_idna, question_bypass, ifindex, flags);
+        if (r == -ENOANO)
+                return sd_varlink_error(link, "io.systemd.Resolve.DNSError", NULL);
```
Maybe, better to introduce a new error code.
`"io.systemd.Resolve.DNSError"` -> `"io.systemd.Resolve.QueryRefused"`
Please also update src/shared/varlink-io.systemd.Resolve.c .

yuwata commented on test/units/TEST-75-RESOLVED.sh:
```diff
@@ -1084,6 +1084,47 @@ testcase_13_varlink_subscribe_dns_configuration() {
+    run dig localhost -t A
+    grep -qF "status: NOERROR" "$RUN_OUT"
+
+    run resolvectl query localhost5
```
Please add more test cases with `resolvectl query --type=AAAA`, `SRV`, `TXT`, and `A`.

muhammadn replied:
added the `resolvectl query --type` test cases for AAAA, SRV, TXT and A.

yuwata reviewed (changes requested):
Looks mostly good. Almost there! But one more round please.
(Making the PR draft again. Please push the ready button when addressed.)

arvidjaar commented:
Unless I miss something, this PR adds static knob to turn off `AAAA` lookups. What was mostly requested since years though (not necessarily from systemd) was the ability to turn off IPv6 when bringing up IPv4 only (VPN) connection on a system that otherwise has IPv6 connectivity. If I do not want to ever resolve IPv6, I would likely simply disable IPv6 and be done with it.

What would really be useful is per-interface option to disable `AAAA` lookups for the traffic via this interface that could be set dynamically when interface comes up.

muhammadn commented:
> Unless I miss something, this PR adds static knob to turn off `AAAA` lookups. What was mostly requested since years though (not necessarily from systemd) was the ability to turn off IPv6 when bringing up IPv4 only (VPN) connection on a system that otherwise has IPv6 connectivity. If I do not want to ever resolve IPv6, I would likely simply disable IPv6 and be done with it.
> 
> What would really be useful is per-interface option to disable `AAAA` lookups for the traffic via this interface that could be set dynamically when interface comes up.

Initially this PR was about refusing `AAAA` queries but then it's too specific. 

But this feature has been changed to to have ability to refuse many types of records, like `SRV, TXT, MX, etc...` and not just `AAAA`. Side effect is that it can also be used to prevent servers from also querying ActiveDirectory (`SRV, TXT`) and follow some DNS policies in the organisation. (example: I don't want my applications connect and send email to any servers - `MX` record)

muhammadn commented:
@yuwata changed as suggested. Also the `noble/*` is due to debconf errors. The `mkosi` error is not related (due to `cryptsetup` and testing-farm is due to `TEST-02-UNITTESTS`.

Not seeing errors with `TEST-75-RESOLVED` so far.

poettering commented on man/resolved.conf.xml:
```diff
@@ -317,6 +317,21 @@
+        <term><varname>RefuseRecordTypes=</varname></term>
+	<listitem><para>Takes a list of DNS record types separated by space.
+	Specified record types in each query will be refused. This option can be specified multiple times.
+	If an empty string is specified, then all previous assignments are cleared.</para>
```
please follow our usual identation, as in the rest of the file.

muhammadn replied:
replaced tabs with spaces (vim can be annoying) :-(

yuwata reviewed (approved):
LGTM.
